src/Env.hs

```diff
@@ -0,0 +1,320 @@
+
+----------------------------------- Parser --------------------------------------------
+
+typeNameParser :: Parser TypeName
+typeNameParser = do
+    i <- identifier
+    return (TypeNameConst i)
+
+typeParser :: Parser Type
+typeParser = do
+    rword "type"
+    pos <- getPosition
+    return Type{ typeName = "type", typePos = pos}
+
+varParser :: Parser Var
+varParser = do
+    i <- identifier
+    return (VarConst i)
+
+typeVarParser :: Parser TypeVar
+typeVarParser = do
+    aps
+    i <- identifier
+    pos <- getPosition
+    return TypeVar { typeVarName =  i, typeVarPos = pos}
+
+yParser, y1, y2 :: Parser Y
+yParser = try y1 <|> y2
+y1 = do
+  i <- varParser
+  return (VarY i)
+y2 = do
+  i <- typeVarParser
+  return (TypeVarY i)
+
+propParser :: Parser Prop
+propParser = do
+    rword "Prop"
+    pos <- getPosition
+    return Prop { propName = "Prop", propPos = pos}
+
+tParser, tConstructorInvokerParser, typeVarParser' :: Parser T
+tParser = try tConstructorInvokerParser <|> typeVarParser'
+tConstructorInvokerParser = do
+    i         <- identifier
+    arguments <- option [] $ parens (argParser `sepBy1` comma) 
+    --try (parens (argParser `sepBy1` comma)) <|> emptyArgList --option [] $ parens (argParser `sepBy1` comma) --try (parens (argParser `sepBy1` comma)) <|> emptyArgList
+    pos <- getPosition
+    return (TConstr (ConstructorInvoker { nameCons = i, argCons = arguments, constructorInvokerPos = pos }))
+typeVarParser' = do
+    i <- typeVarParser
+    return (TTypeVar i)
+
+argParser, varParser', tParser'  :: Parser Arg
+argParser = try tParser' <|> varParser'
+varParser' = do
+     i <- varParser
+     return (AVar i)
+tParser' = do
+    t <- tParser
+    return (AT t)
+
+kParser, kTypeParser, tParser'' :: Parser K
+kParser = try kTypeParser <|> try tParser''
+kTypeParser = do
+     rword "type"
+     pos <- getPosition
+     return (Ktype (Type { typeName = "type", typePos = pos }))
+tParser'' = do
+    t <- tParser
+    return (KT t)
+
+emptyArgList :: Parser [Arg]
+emptyArgList = do
+  lparen
+  rparen
+  return []
+
+----------------------------------- Typechecker aux functions ------------------------------------------
+
+firsts :: [(a, b)] -> [a]
+firsts xs = [x | (x,_) <- xs]
+
+seconds :: [(a, b)] -> [b]
+seconds xs = [x | (_,x) <- xs]
+
+second :: (a, b) -> b
+second (a, b) = b
+
+checkAndGet :: String -> M.Map String v -> SourcePos -> Either String v
+checkAndGet k m pos = case M.lookup k m of
+  Nothing -> Left ("Error in " ++ (sourcePosPretty pos) ++ " : " ++ k ++ " Doesn't exist in the context \n")
+  Just v ->  Right v
+
+lookUpK :: VarEnv -> Arg -> K
+lookUpK e (AT  (TTypeVar t))  = (Ktype ((typeVarMap e) M.! t)) --(Ktype (Type {typeName = "type", typePos = typeVarPos t }))
+lookUpK e (AT  (TConstr t)) =
+        if (nameCons t) `elem` (declaredNames e)
+        then lookUpK e (AVar (VarConst (nameCons t)))
+        else (Ktype (Type { typeName = "type", typePos = constructorInvokerPos t }))
```

Comment:
does this handle the case where the type name is not legitimate -- it is neither a type variable (e.g. `X`) nor a nullary type ctor (e.g. `Set`)? it could be, say, `lsdhfsjdlfsd`, which should throw an error